src/client/api/json_parser/payloads.c

```diff
@@ -182,5 +222,115 @@ cJSON* json_transaction_serialize(transaction_payload_t* tx) {
   return tx_payload;
 }
 
-// TODO
-cJSON* json_tagged_serialize(void* tx) { return NULL; }
+cJSON* json_tagged_serialize(tagged_data_t* tagged_data) {
+  /*
+  {
+    "type": 5,
+    "tag": "484f524e455420464155434554"
+    "data": "494f5441202d2041206e6577206461776e0a436f756e743a203138393030350a5032c2b573"
+  }
+  */
+  cJSON* tagged_data_payload = NULL;
+
+  if (!tagged_data) {
+    printf("[%s:%d] invalid parameter\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  // create payload object
+  if ((tagged_data_payload = cJSON_CreateObject()) == NULL) {
+    printf("[%s:%d] creating tagged data payload object failed\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  // "type": 5,
+  if (!cJSON_AddNumberToObject(tagged_data_payload, JSON_KEY_TYPE, CORE_MESSAGE_PAYLOAD_TAGGED)) {
+    printf("[%s:%d] add payload type failed\n", __func__, __LINE__);
+    cJSON_Delete(tagged_data_payload);
+    return NULL;
+  }
+
+  // tag
+  if (tagged_data->tag) {
+    if (!cJSON_AddStringToObject(tagged_data_payload, JSON_KEY_TAG, (const char* const)tagged_data->tag->data)) {
+      printf("[%s:%d] add tag type failed\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      return NULL;
+    }
+  } else {
+    // add a null tag to tagged data
+    if (!cJSON_AddNullToObject(tagged_data_payload, JSON_KEY_TAG)) {
+      printf("[%s:%d] add null tag payload failed\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      return NULL;
+    }
+  }
+
+  // data
+  if (tagged_data->data) {
+    char* data_str = malloc(BIN_TO_HEX_STR_BYTES(tagged_data->data->len));
+    if (!data_str) {
+      printf("[%s:%d] OOM\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      return NULL;
+    }
+    if (bin_2_hex(tagged_data->data->data, tagged_data->data->len, data_str, sizeof(data_str)) != 0) {
+      printf("[%s:%d] bin to hex data conversion failed\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      free(data_str);
+      return NULL;
+    }
+    if (!cJSON_AddStringToObject(tagged_data_payload, JSON_KEY_DATA, data_str)) {
+      printf("[%s:%d] add data type failed\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      free(data_str);
+      return NULL;
+    }
+    free(data_str);
+  } else {
+    // add a null data to tagged data
+    if (!cJSON_AddNullToObject(tagged_data_payload, JSON_KEY_DATA)) {
+      printf("[%s:%d] add null data payload failed\n", __func__, __LINE__);
+      cJSON_Delete(tagged_data_payload);
+      return NULL;
+    }
+  }
+
+  return tagged_data_payload;
+}
+
+int json_tagged_deserialize(cJSON* payload, tagged_data_t** tagged_data) {
+  if (!payload) {
```

Comment:
`tagged_data` should be checked too.
```suggestion
  if (!payload || *tagged_data) {
```

Reply:
Fixed. Thanks!

Reply:
https://github.com/iotaledger/iota.c/pull/305/commits/7527ace33618e491e9af74f462f24aa2e01c2204 doesn't fix the problem. 
the value `*tagged_data` should be checked as the suggestion above, not the address of the value `tagged_data`.
same as line [90](https://github.com/iotaledger/iota.c/pull/305/files#diff-f16bd3240fb36df6ddc7b455d53f2e34fbe0330fb3f54bd4a14d3cc1b8e00012R90)
it should be 
```
if (!essence_payload || *tagged_data) {
```

Reply:
Thank you for the explanation. I got it where it was a problem. It's fixed now.